### PR TITLE
Add unit tests for TensorRT integration and fix some bugs

### PR DESCRIPTION
**Pull request by Caenorst**
## Description ##
TensorRT integration lacked of unit tests, we instead relied on output comparison of a full network which is not very pertinent, difficult to find the tolerance and not very helpful if it fails.

This PR have two purposes:
 1) Add unit test for all operations:
As we only partition subgraph of at least 2 Ops we always append an identity to each output. we then compare to MXNet, using both TRT FP32 and FP16 computation.
2) Fix a bunch of edge cases bugs that have been exposed by the unit tests:
  - NHWC is currently not compatible with TensorRT
  - Pooling is currently not compatible with count_include_pad and only compatible with pooling_convention Valid
  - FullyConnected without bias cannot be converted to MatMul (as there should be a transpose before) for the moment we decided to remove compatibility, it is quite rare to do FullyConnected without bias anyway
  - Concat is not compatible if the concatenation axis is the batch axis
  - The dropout have to be only enabled on training mode (act as identity)
  - BatchNorm with fix_gamma can have gamma != 1. so we force the value to be 1. when loading it.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##
- the test of output comparison in a full networks should probably be replaced by an accuracy comparison over a full dataset

### CODE REVIEW

**Comment by anirudhacharya** [comment]
@mxnet-label-bot add [pr-awaiting-review]

**Comment by Caenorst** [comment]
@KellenSunderland @haohuanw

**Comment by KellenSunderland** [comment]
Looks like CI caught a few issues.  For example http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Funix-gpu/detail/PR-15399/1/pipeline seems like it should be relevant to this PR.  I'll have a look to see if there's anything else that jumps out at me.

File: src/operator/subgraph/tensorrt/nnvm_to_onnx.cc [hunk 1]
```diff
@@ -157,6 +157,12 @@ std::string ConvertNnvmGraphToOnnx(
   return serialized_onnx_graph;
 }
 
+void ConvertIdentity(NodeProto* node_proto, const NodeAttrs& attrs,
```
Comment (KellenSunderland) [inline src/operator/subgraph/tensorrt/nnvm_to_onnx.cc]: Any idea if TRT actually optimizes this out?  I've seen this in a few prod services :-/

File: src/operator/subgraph/tensorrt/tensorrt-inl.h [hunk 2]
```diff
@@ -109,13 +111,70 @@ class TensorrtSelector : public SubgraphSelector {
...
+          return false;
+      }
+    }
+
+    if (op_name == "Concat") {
+      const auto& param = nnvm::get<ConcatParam>(n.attrs.parsed);
+      return (param.dim != 0);
+    }
+
+    if (op_name == "Dropout") {
```
Comment (KellenSunderland) [inline src/operator/subgraph/tensorrt/tensorrt-inl.h]: Again, will TensorRT optimize this out?  We don't want it at inference time right?

File: src/operator/subgraph/tensorrt/tensorrt-inl.h [hunk 3]
```diff
@@ -180,6 +253,17 @@ class TensorrtProperty : public SubgraphProperty {
     n->attrs.name = "TensorRT" + std::to_string(subgraph_id);
     n->attrs.op = Op::Get("_TensorRT");
     CHECK(n->attrs.op);
+    // prevent using Gamma value if using fix_gamma on BatchNorm
+    DFSVisit(new_sym.outputs, [&n](const nnvm::NodePtr& node) {
+      if (node->op() == Op::Get("BatchNorm")) {
```
Comment (KellenSunderland) [inline src/operator/subgraph/tensorrt/tensorrt-inl.h]: Why not just check for FixGamma = true during nnvm -> Onnx conversion and set gamma to 0 if it's true?

File: tests/python/gpu/test_tensorrt.py [hunk 4]
```diff
@@ -0,0 +1,437 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (KellenSunderland) [inline tests/python/gpu/test_tensorrt.py]: Super helpful tests, thanks Clement.  It's ok to merge in a subset of these at a time, but I'd try to avoid pushing commented out code / too many todos upstream.

**Comment by karan6181** [comment]
@Caenorst Could you please address the review comments? Thanks!

**Comment by piyushghai** [comment]
@Caenorst Gentle ping...

File: src/operator/subgraph/tensorrt/tensorrt-inl.h (hunk 2 above)
Comment (Caenorst) [reply src/operator/subgraph/tensorrt/tensorrt-inl.h]: Dropout have always been seen as identity function in MXNet-TensorRT integration so I don't see any changement on this, regarding to whether or not identity is actually doing a copy or not I'm not quite sure, here is the onnx-tensorrt conversion: https://github.com/onnx/onnx-tensorrt/blob/0ab159579551cabfa05fd66f338357f116e96835/trt_utils.hpp#L169-L180

Comment (KellenSunderland) [reply src/operator/subgraph/tensorrt/tensorrt-inl.h]: Ok, non-blocking comment for this PR.  I'm just thinking about adding a warning in the future if people are using TRT with operations that don't make sense at inference time (Dropout, Ident, Empty Concats or Copies, etc.)

File: src/operator/subgraph/tensorrt/tensorrt-inl.h (hunk 3 above)
Comment (KellenSunderland) [reply src/operator/subgraph/tensorrt/tensorrt-inl.h]: I think this one should be handled upstream.  Have a look at https://github.com/apache/incubator-mxnet/commit/6b22aa455e562db0e2744e94dc6a583191081b05 and let us know if you'd rather use this fix or if ours looks ok.

File: src/operator/subgraph/tensorrt/nnvm_to_onnx.cc (hunk 1 above)
Comment (Caenorst) [reply src/operator/subgraph/tensorrt/nnvm_to_onnx.cc]: I believe this should be optimized by ONNX-TRT

File: src/operator/subgraph/tensorrt/tensorrt-inl.h (hunk 3 above)
Comment (Caenorst) [reply src/operator/subgraph/tensorrt/tensorrt-inl.h]: I think solution 6b22aa4 looks better, I removed mine

**Comment by Caenorst** [comment]
I don't understand the error on windows-gpu, it doesn't seems related to my modifications...

**Comment by Caenorst** [comment]
@KellenSunderland can we merge it ? (I did a bunch of modifications since the last review that you may wanna review too)

File: tests/python/gpu/test_tensorrt.py (hunk 4 above)
Comment (Caenorst) [reply tests/python/gpu/test_tensorrt.py]: I removed the commented out unit tests
